Add docs/requirements.md to outline CLI functionality

satukoskinen opened the pull request:
This document outlines the to-be-implemented functionality for `hp-cli`.

Closes #1

satukoskinen committed:
Add docs/requirements.md to outline CLI functionality

Signed-off-by: Satu Koskinen <satu.a.koskinen@gmail.com>

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+            <timestamp>_gps.json
+            ...
+        depth/
+            <timestamp>_depth.csv
```
Depth recording should only have one .csv file.

satukoskinen replied:
I think we should have the option to merge multiple depth csv files in case the service has restarted during the deployment and started writing to a new file.

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+            <timestamp>_depth.csv
+            ...
+        log/
+            <timestamp>_journalctl.txt
```
Log should only have one journalctl log.

juliuskoskela commented on docs/requirements.md:
````diff
@@ -0,0 +1,146 @@
+
+```
+/output/
+    <deployment-timestamp>/
````
What is `deployment`? I think we can only identify deployments on a single device by timestamp.

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+
+--no-metadata               Do not prompt the user for submitting metadata about a deployment.
+
+--import-path PATH          Alternative path on the device file system from which data should
```
There should be no need for alternative path.

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+                            Resulting file will have as its name <timestamp>_audio.wav
+                            where the timestamp is the timestamp of the first wav file.
+
+--no-metadata               Do not prompt the user for submitting metadata about a deployment.
```
Maybe instead of this we would have metadata fields as options which would then forego the prompt.

I was also thinking about the categorization method. Maybe tags would be the best? You could put `--tags ["andenes", "norway", "whale2sea"]` for example.

```
--name
--tags
--team
```

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+DEVICE                      Path to USB mass storage or SD card where data will
+                            be imported from.
+
+--output-path OUTPUT        Path to where the directory for imported data
```
I don't think this should be an option.

satukoskinen replied:
I think output path should of course be an option. How do we know where the user wants to import the data to?

juliuskoskela replied:
I mean it doesn't need a flag, because it's a required argument.

```
hp-cli import /device/path /output/path --some-flag true

# as opposed to...

hp-cli import /device/path -o /output/path --some-flag true
```

juliuskoskela commented on docs/requirements.md:
````diff
@@ -0,0 +1,146 @@
+Introduces the following command:
+
+```sh
+hp-cli import DEVICE --output-path OUTPUT [options]
````
I think output path shouldn't be an option, but rather an argument.

```
hp-cli import [DEVICE] [OUTPUT] [options]
```

juliuskoskela commented on docs/requirements.md:
```diff
@@ -0,0 +1,146 @@
+
+At the beginning of the import, the CLI lists the deployments it has detected. For each deployment, the size of the deployment folder is printed. A progress bar shows the progress of the import.
+
+Upon successful import of a single deployment, a metadata file is created. CLI asks user to fill in optional information unless ``--no-metadata`` is supplied. 
```
Maybe we rather require the metadata, but can pass the fields as optional arguments.

juliuskoskela commented on docs/requirements.md:
````diff
@@ -0,0 +1,146 @@
+#### 1.2 Imported Data Directory Structure
+
+```
+<timestamp>-<device_name>/
+    meta.json
+    audio/
+        <timestamp>_audio.wav
+        ...
+    gps/
+        <timestamp>_gps.json
+        ...
+    depth/
+        <timestamp>_depth.csv
+        ...
+    log/
+        <timestamp>_journalctl.txt
+        ...
+```
````
Check changes to depth and log

juliuskoskela commented on docs/requirements.md:
````diff
@@ -0,0 +1,146 @@
+
+When importing a dataset from a deployment, the CLI interface should ask the user to fill out optional deployment info. Some of the fields are inferred from the data itself. This data will be saved as `meta.json` file.
+
+```rs
+struct DeploymentInfo {
+    tags: Option<Vec<String>>,
+    notes: Option<String>,
+    start: chrono::DateTime<Utc>, // Inferred from the timestamp of first audio file.
+    end: chrono::DateTime<Utc>, // Inferred from the timestamp of last audio file.
+}
+```
````
Let's think about the metadata... Maybe add name as well?

juliuskoskela reviewed (changes requested):
Maybe add the following topic:

- Checks. Import should check for data integrity. Report if there are unnecessary files and analyze if the audio has dropped between patches.

satukoskinen committed:
Update docs/requirements.md

Signed-off-by: Satu Koskinen <satu.a.koskinen@gmail.com>

juliuskoskela committed:
Update requirements.md

- Improve command descriptions to match the rust library `clap`.
- Add a section on correct timestamp.

Signed-off-by: Julius Koskela <me@juliuskoskela.dev>

juliuskoskela committed:
Update requirements.md

Line wrapping

juliuskoskela committed:
Update requirements.md

Change Utc to Local time in an example

juliuskoskela committed:
Update requirements.md

- Improve on imported driectory structure.

Signed-off-by: Julius Koskela <me@juliuskoskela.dev>

juliuskoskela reviewed (commented):
I think this version is good enough for review by the develpers.
